common/eth2/src/types.rs

```diff
@@ -1008,15 +1009,40 @@ impl FromStr for Accept {
     type Err = String;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            "application/octet-stream" => Ok(Accept::Ssz),
-            "application/json" => Ok(Accept::Json),
-            "*/*" => Ok(Accept::Any),
-            _ => Err("accept header cannot be parsed.".to_string()),
+        let mimes = parse_accept(s);
+
+        // [q-factor weighting]: https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.2
+        let mut highest_mime: Option<Mime> = None;
+        let mut max_pref = 0f32;
+        for m in mimes {
+            let pref = m
+                .get_param("q")
+                .map_or(1f32, |n| n.as_ref().parse::<f32>().unwrap_or(0f32));
+            if pref > max_pref {
+                highest_mime = Some(m);
+                max_pref = pref;
+            }
         }
+
+        highest_mime.map_or(
+            Err("accept header cannot be parsed.".to_string()),
+            |m| match (m.type_(), m.subtype()) {
+                (APPLICATION, OCTET_STREAM) => Ok(Accept::Ssz),
+                (APPLICATION, JSON) => Ok(Accept::Json),
+                (STAR, STAR) => Ok(Accept::Any),
+                _ => Err("accept header is not supported".to_string()),
+            },
+        )
     }
 }
 
+fn parse_accept(accept: &str) -> Vec<Mime> {
+    accept
+        .split(',')
```

Comment:
Splitting on commas is not _technically correct_ because the header could contain quoted strings which contain commas. However it seems that no correct implementation of Accept header parsing exists in Rust, so we should be practical and merge something that works 99% of the time.

I'm just noting this, in case you or anyone else reading this feels like contributing a correct impl to the Rust ecosystem.

Reply:
I asked nicely and now this library has a parser for this you can use: https://github.com/picoHz/mediatype/issues/4 https://docs.rs/mediatype/latest/mediatype/struct.MediaTypeList.html

Reply:
Thank you @glasser that's awesome!